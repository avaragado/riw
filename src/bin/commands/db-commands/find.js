// @flow

import type yargs from 'yargs';

import outdent from 'outdent';
import chalk from 'chalk';
import pick from 'ramda/src/pick';

import { createHandlerWithRIW } from '../../utils';

export const command = 'find';
export const desc = 'Find translation database entries matching options';

const here = `db ${command}`;

export const builder = (yyargs: yargs.Argv) => yyargs
    .usage(outdent`
        $0 ${here}
            [--defaultMessage <sourceString>]
            [--description <description>]
            [--locale <localeID]
            [--translation <localeString>]
            [--json]

        Outputs translation database entries matching all of the options supplied.
        Omitted options match any value.

        Each database entry comprises the string in the default locale, the description
        for that string, the locale of the translated string, and the translated string itself.
    `)
    .example(
        outdent`
            $0 ${here}
                --defaultMessage "Hello {name}"

        `,
        'Finds all database entries for the defaultMessage "Hello {name}", and outputs in readable form.',
    )
    .example(
        outdent`
            $0 ${here}
                --locale fr-fr
                --json

        `,
        'Finds all database entries for translations into fr-fr, and outputs as JSON',
    )
    .example(
        outdent`
            $0 ${here}
                --description _
                --locale pt-br

        `,
        'Finds all database entries for translations into pt-br with the default description, _, and outputs in readable form ',
    )
    .options({
        defaultMessage: {
            alias: 'm',
            string: true,
            group: 'Command options',
            desc: outdent`
                Output entries matching the string being translated.
                Omit to match any string.
            `,
        },

        description: {
            alias: 'd',
            string: true,
            group: 'Command options',
            desc: outdent`
                Output entries with this description.
                Omit to match any description.
            `,
        },

        locale: {
            alias: 'l',
            string: true,
            group: 'Command options',
            desc: outdent`
                Output entries with this locale.
                Omit to match any locale.
            `,
        },

        translation: {
            alias: 't',
            string: true,
            group: 'Command options',
            desc: outdent`
                Output entries with this translated string.
                Omit to match any translated string.
            `,
        },

        json: {
            boolean: true,
            group: 'Command options',
            desc: outdent`
                Output entries in JSON format as an array of
                [defaultMessage, description, locale, translation] arrays.
                Omit to output in human-readable form.
            `,
        },
    });

export const handler = createHandlerWithRIW((riw: RIW, argv: yargs.Argv) => {
    const opt: RIWCLIOptDBFind = {
        match: pick(['defaultMessage', 'description', 'locale', 'translation'], argv),
    };

    const quads = riw.findTranslations(opt);

    if (argv.json) {
        console.log(JSON.stringify(quads, null, 4));

    } else {
        if (quads.length > 0) {
            console.log(
                chalk.bold('Matches [%d]:\n'),
                quads.length,
            );
        } else {
            console.log(
                chalk.bold('No matches'),
            );
        }

        quads.forEach((quad) => {
            console.log(
                ' -',
                chalk.bold.blue(quad[0]),
                chalk.dim(quad[1]),
                '\n',
                ' ',
                chalk.green(quad[2]),
                chalk.bold.green(quad[3]),
                '\n',
            );
        });
    }

});
