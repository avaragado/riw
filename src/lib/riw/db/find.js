// @flow

import compose from 'ramda/src/compose';

import dbRead from './rw/dbRead';
import arquadFromDB from './transform/arquadFromDB';
import arquadFindFromQuadAr from './transform/arquadFindFromQuadAr';

export default (config: RIWConfig) => (opt?: RIWCLIOptDBFind) => compose(
    arquadFindFromQuadAr(config, opt),
    arquadFromDB,
    dbRead(config),
)();
