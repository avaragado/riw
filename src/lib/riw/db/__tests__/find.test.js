// @flow

import mock from 'mock-fs';

import find from '../find';
import cfgBase from '../../__tests__/helpers/dummyConfig';

const cfg = {
    ...cfgBase,
    translationsDatabaseFile: 'db.json',
};

const db: RIWDB = {
    version: 1,
    data: {
        'one': {
            'desc1': {
                'aa-aa': '[aa-aa]1 one',
                'bb-bb': '[bb-bb]1 one',
            },
            'desc2': {
                'aa-aa': '[aa-aa]2 one',
                'bb-bb': '[bb-bb]2 one',
            },
        },
        'two': {
            'desc1': {
                'aa-aa': '[aa-aa]1 two',
                'bb-bb': '[bb-bb]1 two',
            },
        },
        'three': {
            'desc1': {
                'aa-aa': '[aa-aa]1 three',
                'cc-cc': '[cc-cc]1 three',
            },
        },
    },
};

const jsonDB = JSON.stringify(db, null, 4);

type Fixture = {
    name: string,
    opt: RIWCLIOptDBFind,
};

const fixtures: Fixture[] = [
    {
        name: '01 defaultMessage match',
        opt: {
            match: {
                defaultMessage: 'two',
            },
        },
    },

    {
        name: '02 description match',
        opt: {
            match: {
                description: 'desc1',
            },
        },
    },

    {
        name: '03 locale match',
        opt: {
            match: {
                locale: 'bb-bb',
            },
        },
    },

    {
        name: '04 translation match',
        opt: {
            match: {
                translation: '[cc-cc]1 three',
            },
        },
    },
];

describe('lib/riw/db/find', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                [cfg.translationsDatabaseFile]: jsonDB,
            });

            const received = find(cfg)(fixture.opt);

            mock.restore();

            expect(received).toMatchSnapshot();
        });
    });
});
