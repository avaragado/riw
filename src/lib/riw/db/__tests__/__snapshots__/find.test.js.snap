// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/db/find 01 defaultMessage match 1`] = `
Array [
  Array [
    "two",
    "desc1",
    "aa-aa",
    "[aa-aa]1 two",
  ],
  Array [
    "two",
    "desc1",
    "bb-bb",
    "[bb-bb]1 two",
  ],
]
`;

exports[`lib/riw/db/find 02 description match 1`] = `
Array [
  Array [
    "one",
    "desc1",
    "aa-aa",
    "[aa-aa]1 one",
  ],
  Array [
    "one",
    "desc1",
    "bb-bb",
    "[bb-bb]1 one",
  ],
  Array [
    "two",
    "desc1",
    "aa-aa",
    "[aa-aa]1 two",
  ],
  Array [
    "two",
    "desc1",
    "bb-bb",
    "[bb-bb]1 two",
  ],
  Array [
    "three",
    "desc1",
    "aa-aa",
    "[aa-aa]1 three",
  ],
  Array [
    "three",
    "desc1",
    "cc-cc",
    "[cc-cc]1 three",
  ],
]
`;

exports[`lib/riw/db/find 03 locale match 1`] = `
Array [
  Array [
    "one",
    "desc1",
    "bb-bb",
    "[bb-bb]1 one",
  ],
  Array [
    "one",
    "desc2",
    "bb-bb",
    "[bb-bb]2 one",
  ],
  Array [
    "two",
    "desc1",
    "bb-bb",
    "[bb-bb]1 two",
  ],
]
`;

exports[`lib/riw/db/find 04 translation match 1`] = `
Array [
  Array [
    "three",
    "desc1",
    "cc-cc",
    "[cc-cc]1 three",
  ],
]
`;
